pca_cca.go

```diff
@@ -13,51 +13,89 @@ import (
 	"github.com/gonum/matrix/mat64"
 )
 
-// PrincipalComponents returns the principal component vectors and the variances
-// of the principal component scores of the input data which is represented as
-// an n×d matrix a where each row is an observation and each column is a
-// variable.
+// PC is the result of a principal components analysis. The results of a principal
+// components analysis held by a PC are only valid after a call to PrincipalComponents
+// has returned true.
+type PC struct {
+	n, d    int
+	weights []float64
+	svd     *mat64.SVD
+	ok      bool
+}
+
+// PrincipalComponents performs a weighted principal components analysis on the
+// matrix of the input data which is represented as an n×d matrix a where each
+// row is an observation and each column is a variable.
 //
 // PrincipalComponents centers the variables but does not scale the variance.
 //
 // The slice weights is used to weight the observations. If weights is nil, each
 // weight is considered to have a value of one, otherwise the length of weights
 // must match the number of observations or PrincipalComponents will panic.
 //
-// On successful completion, the principal component vectors are returned in the
-// columns of vecs as a d×min(n,d) matrix, and the column variances of the
-// principal component scores, b * vecs where b is a with centered columns, are
-// returned in vars as a min(n,d)-long slice in descending sort order.
-//
-// On failure, vecs and vars are returned nil and ok is returned false.
-func PrincipalComponents(a mat64.Matrix, weights []float64) (vecs *mat64.Dense, vars []float64, ok bool) {
-	n, d := a.Dims()
-	if weights != nil && len(weights) != n {
+// PrincipleComponents returns whether the analysis was successful.
+func (c *PC) PrincipalComponents(a mat64.Matrix, weights []float64) (ok bool) {
+	c.n, c.d = a.Dims()
+	if weights != nil && len(weights) != c.n {
 		panic("stat: len(weights) != observations")
 	}
 
-	svd, ok := svdFactorizeCentered(a, weights)
-	if !ok {
-		return nil, nil, false
+	c.svd, c.ok = svdFactorizeCentered(c.svd, a, weights)
+	if c.ok {
+		c.weights = append(c.weights[:0], weights...)
+	}
+	return c.ok
+}
+
+// Vectors returns the component direction vectors of a principal components
+// analysis. The vectors are returned in the columns of a d×n matrix.
+// If dst is not nil it must either be zero-sized or be a d×n matrix. dst will
```

Comment:
Also yeah as far as I can tell from the current implementation it should be `d x min(d,n)`, however it really should be `d x min(d, n-1)` because due to the centering done, the `n` th singular value will always be zero when `d < n`. I'm not sure how to implement this though other than just checking for it and removing the last column if necessary -- which seems ugly, I guess just leave the implementation as it is but a note in the documentation is probably appropriate. End of the day having an extra useless principal component vector in some cases won't do any harm I suppose.